File: exercises/py110/easy2/halvsies.py
'''
PEDAC
Input: a list
Output: a list that contains two elements, both of which are lists.

Rules
Explicit:
- Put the first half of the original list in the first element of the return value
- Put the second half of the original list in the second element of the return value
- If the original list contains an odd number of elements, place the middle element in the first list 
Implicit:
- 

Questions:
-What if the original list is empty?
    -Both lists in the returning list will be empty lists

Data Structures:
-list 

Algorithm: 
    - initialize a "container_list" as an empty list []
    - initialize "first_element" as an empty list []
    - initialize "second_element" as an empty list []

    - if the length of the original list "container_list" is 0, 
        - container_list.append(first_element), container_list.append(second_element)
        - return container_list 
    - get the middle index of the original list 
        - "middle_index" = len(lst) / 2 
        - append first half of the original list to "first_element"
        - append second half of the original list to "second_element"
        - append first and second element to container_list
    - return container_list 
'''

def halvsies(lst):
    # 6 : 3, 5 : 3 / 4 : 2, 3 : 2
    middle_index = (len(lst) + 1) // 2
    first_half = lst[:middle_index]
    second_half = lst[middle_index:]
   
    return [first_half, second_half]
    
# All of these examples should print True
print(halvsies([1, 2, 3, 4]) == [[1, 2], [3, 4]])
print(halvsies([1, 5, 2, 4, 3]) == [[1, 5, 2], [4, 3]])
print(halvsies([5]) == [[5], []])
print(halvsies([]) == [[], []])

File: exercises/LS215/easy4/halvsies.js

/*

=== Problem ===
Input: 1D array
Output: array with two nested array elements
        - first arr:  first half of the original arr
        - second arr: second half of the original arr

Rules:
- If the original array contains an odd number of elements, place the middle element in the first half array.

=== Data Structures ===
- array
- int

=== Brainstorm ===
- get the middle index:
  - arr.length / 2

=== Algorithm === 


  [1, 2, 3, 4] => [[1, 2], [3, 4]]
  middleIdx = 2
  firstHalf = nums(0, 2) => [1, 2]
  secondHalf = nums(2)   => [3, 4]
  return  [[1, 2], [3, 4]]

  [1, 5, 2, 4, 3] => [[1, 5, 2], [4, 3]]
  middleIdx = Math.floor(2.5) = 2
  arr has odd length, middleIdx += 1, 
    middleIdx = 3
  firstHalf = nums(0, 3) => [1, 5, 2]
  secondHalf = nums(3)   => [4, 3]
  returns [[1, 5, 2], [4, 3]]


- initialize a variable `middleIdx` to length of the array divided by 2
- Math.floor(nums.length / 2)
- convert the number to int using Math.floor() method

- if original arr has an odd length, add 1 to `middleIdx`
- initialize an arr `firstHalf` to original sliced with (0, middle)
  - nums.slice(0, middle)
- initialize an arr `secondHalf` to original sliced with (middle)
  - nums.slice(middle)
- return [firstHalf, secondHalf
  - array with two nested array elements
*/
function halvsies(nums) {
  let middleIdx = Math.floor(nums.length / 2);
  if (isOddArr(nums)) middleIdx += 1;

  // console.log(middleIdx); // 2, 3, 1, 0
  let firstHalf = nums.slice(0, middleIdx);
  let secondHalf = nums.slice(middleIdx);
  
  return [firstHalf, secondHalf];
}

const isOddArr = (arr) => arr.length % 2 === 1;

console.log(halvsies([1, 2, 3, 4]));       // [[1, 2], [3, 4]]
console.log(halvsies([1, 5, 2, 4, 3]));    // [[1, 5, 2], [4, 3]]
console.log(halvsies([5]));                // [[5], []]
console.log(halvsies([]));                 // [[], []]